### PR TITLE
OMX AAC audio decoder element

### PR DESCRIPTION
**Pull request by joselopez10**

### CODE REVIEW

File: omx/gstomxaacdec.c [hunk 1]
```diff
@@ -22,12 +22,14 @@
...
 
 #include <gst/gst.h>
-
 #include "gstomxaacdec.h"
 
 GST_DEBUG_CATEGORY_STATIC (gst_omx_aac_dec_debug_category);
 #define GST_CAT_DEFAULT gst_omx_aac_dec_debug_category
 
+#define OMX_AACDEC_INPUT_PORT 0
+#define OMX_AACDEC_OUTPUT_PORT 1
```
Comment (cgomezv) [inline omx/gstomxaacdec.c]: We could move these two definitions to the helper defs for TI DM81xx section in gstomx.h, and modify the name to GST_OMX_AAC_DEC_ to follow the pattern.

File: omx/gstomxaacdec.c [hunk 2]
```diff
@@ -62,10 +64,10 @@ gst_omx_aac_dec_class_init (GstOMXAACDecClass * klass)
       GST_DEBUG_FUNCPTR (gst_omx_aac_dec_get_channel_positions);
 
   audiodec_class->cdata.default_sink_template_caps = "audio/mpeg, "
-      "mpegversion=(int){2, 4}, "
-      "stream-format=(string) { raw, adts, adif, loas }, "
+      "mpegversion=(int){ 2, 4 }, "
+      "stream-format=(string) { raw, adts, adif }, "
       "rate=(int)[8000,48000], "
-      "channels=(int)[1,9], " "framed=(boolean) true";
+      "channels=(int)[1,2], " "framed=(boolean) true";
```
Comment (cgomezv) [inline omx/gstomxaacdec.c:67]: The AAC encoder for dm81xx support up to 5 channels, does the decoder support only 2 channels?

File: omx/gstomxaacdec.c [hunk 3]
```diff
@@ -79,7 +81,7 @@ gst_omx_aac_dec_class_init (GstOMXAACDecClass * klass)
 static void
 gst_omx_aac_dec_init (GstOMXAACDec * self)
 {
-  /* FIXME: Other values exist too! */
+  /* NOTE: If necessary adjust to match TI aacdec setup */
   self->spf = 1024;
```
Comment (cgomezv) [inline omx/gstomxaacdec.c]: We could create a definition for this number, it's similar to the GST_OMX_AAC_ENC_INBUF_NSAMPLES for the AAC encoder case.

File: omx/gstomxaacdec.c [hunk 4]
```diff
@@ -96,70 +98,113 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
...
 
   gst_omx_port_get_port_definition (port, &port_def);
-  port_def.format.audio.eEncoding = OMX_AUDIO_CodingAAC;
+
+  if (port->index == OMX_AACDEC_INPUT_PORT) {
+    port_def.nPortIndex = port->index;
+    port_def.eDir = OMX_DirInput;
+    port_def.nBufferCountActual = 1;
+    port_def.nBufferCountMin = 1;
+    port_def.nBufferSize = 4096;
```
Comment (cgomezv) [inline omx/gstomxaacdec.c]: We could create a definition for this number as well

File: omx/gstomxaacdec.c [hunk 5]
```diff
@@ -96,70 +98,113 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
...
+    port_def.format.audio.eEncoding = OMX_AUDIO_CodingAAC;
+    port_def.format.audio.bFlagErrorConcealment = OMX_FALSE;
+    GST_DEBUG_OBJECT (self, "Updating input port definition");
+
+  } else if (port->index == OMX_AACDEC_OUTPUT_PORT) {
+    port_def.nPortIndex = port->index;
+    port_def.eDir = OMX_DirOutput;
+    port_def.nBufferCountActual = 1;
+    port_def.nBufferCountMin = 1;
+    port_def.nBufferSize = 4608; // port_def.nBufferSize = 8192 // TI setup
```
Comment (cgomezv) [inline omx/gstomxaacdec.c]: Why 4608 and not 8192? Maybe the comment is confusing.

File: omx/gstomxaudiodec.c [hunk 6]
```diff
@@ -423,6 +415,9 @@ gst_omx_audio_dec_loop (GstOMXAudioDec * self)
       gst_audio_get_channel_reorder_map (pcm_param.nChannels, self->position,
           omx_position, self->reorder_map);
 
+  GST_DEBUG_OBJECT (self, "PCM params: eNumData %d, eEndian %d, nBitPerSample %d",
+    pcm_param.eNumData, pcm_param.eEndian, pcm_param.nBitPerSample);
```
Comment (cgomezv) [inline omx/gstomxaudiodec.c]: Cast the arguments with type 'OMX_U32' to 'int' to avoid having compilation warnings.

File: omx/gstomxaudiodec.c [hunk 7]
```diff
@@ -345,19 +337,19 @@ gst_omx_audio_dec_loop (GstOMXAudioDec * self)
     /* Just update caps */
     GST_AUDIO_DECODER_STREAM_LOCK (self);
 
-    gst_omx_port_get_port_definition (port, &port_def);
-    g_assert (port_def.format.audio.eEncoding == OMX_AUDIO_CodingPCM);
```
Comment (cgomezv) [inline omx/gstomxaudiodec.c:349]: The port_def variable is unused with this change, so please remove it from line 309.

File: omx/gstomxaudiodec.c [hunk 8]
```diff
@@ -867,29 +862,24 @@ gst_omx_audio_dec_set_format (GstAudioDecoder * decoder, GstCaps * caps)
...
+      return FALSE;
+    }
+  }
 
-  /* Get codec data from caps */
+  /* Disable codec data object usage */
   gst_buffer_replace (&self->codec_data, NULL);
-  s = gst_caps_get_structure (caps, 0);
-  codec_data = gst_structure_get_value (s, "codec_data");
-  if (codec_data) {
```
Comment (cgomezv) [inline omx/gstomxaudiodec.c:886]: With this change 's' and 'codec_data' variables become unused, so please remove them in lines 792 and 793.

File: omx/gstomxaudiodec.c [hunk 9]
```diff
@@ -970,6 +977,7 @@ gst_omx_audio_dec_set_format (GstAudioDecoder * decoder, GstCaps * caps)
 
   self->downstream_flow_ret = GST_FLOW_OK;
 
+  GST_DEBUG_OBJECT (self, "Decoder format set");
```
Comment (cgomezv) [inline omx/gstomxaudiodec.c]: Is this debug comment really necessary?

File: omx/gstomxaacdec.c (hunk 2 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c:67]: Yes, the aac decoder supports only standard stereo 1/2 channels.

File: omx/gstomxaudiodec.c (hunk 7 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:349]: 66cee90a7d9904dfc8155591da324492d4f520ef

File: omx/gstomxaudiodec.c (hunk 8 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:886]: 66cee90a7d9904dfc8155591da324492d4f520ef

File: omx/gstomxaacdec.c (hunk 1 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: 66cee90a7d9904dfc8155591da324492d4f520ef

File: omx/gstomxaacdec.c (hunk 3 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: 66cee90a7d9904dfc8155591da324492d4f520ef

File: omx/gstomxaacdec.c (hunk 4 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: 66cee90a7d9904dfc8155591da324492d4f520ef

File: omx/gstomxaacdec.c (hunk 5 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: 66cee90a7d9904dfc8155591da324492d4f520ef

File: omx/gstomxaudiodec.c (hunk 6 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c]: 66cee90a7d9904dfc8155591da324492d4f520ef

**Review by eguzmanr — APPROVED** [review]

File: omx/gstomxaacdec.c [hunk 10]
```diff
@@ -96,70 +94,114 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
   const gchar *stream_format;
 
   gst_omx_port_get_port_definition (port, &port_def);
-  port_def.format.audio.eEncoding = OMX_AUDIO_CodingAAC;
+
+  if (port->index == GST_OMX_AAC_DEC_INPUT_PORT) {
+    port_def.nPortIndex = port->index;
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: There are many replicated port definitions. I think its better if you factor them out and inside the `ifs` you only set the ones that are different.

File: omx/gstomxaacdec.c [hunk 11]
```diff
@@ -96,70 +94,114 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
...
+    aac_param.eAACProfile = OMX_AUDIO_AACObjectLC;
+    aac_param.nChannels = channels;
+    aac_param.nSampleRate = rate;
+    dec->pcm_param.nSamplingRate = rate;
+
+    if (mpegversion == 2)
+      aac_param.eAACStreamFormat = OMX_AUDIO_AACStreamFormatMP2ADTS;
+    else if (strcmp (stream_format, "adts") == 0)
+      aac_param.eAACStreamFormat = OMX_AUDIO_AACStreamFormatMP4ADTS;
+    else if (strcmp (stream_format, "loas") == 0)
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: You removed this one from the caps.

File: omx/gstomxaacdec.c [hunk 12]
```diff
@@ -96,70 +94,114 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
...
-        gst_omx_error_to_string (err), err);
+        "Failed to update port %d definition with AAC format: %s (0x%08x)",
+        port->index, gst_omx_error_to_string (err), err);
     return FALSE;
   }
 
-  GST_OMX_INIT_STRUCT (&aac_param);
-  aac_param.nPortIndex = port->index;
+  /* Set AAC params using sink pad caps */
+  if (port->index == GST_OMX_AAC_DEC_INPUT_PORT && caps != NULL ) {
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: If `caps == NULL` shouldn't you return an error?  If there are no caps, then the encoder parameters will not be able to be set.

File: omx/gstomxaudiodec.c [hunk 13]
```diff
@@ -252,11 +252,6 @@ gst_omx_audio_dec_change_state (GstElement * element, GstStateChange transition)
     case GST_STATE_CHANGE_PAUSED_TO_PLAYING:
       break;
     case GST_STATE_CHANGE_PAUSED_TO_READY:
-      if (self->dec_in_port)
```
Comment (michaelgruner) [inline omx/gstomxaudiodec.c:255]: Why is this no longer required?

File: omx/gstomxaudiodec.c [hunk 14]
```diff
@@ -279,9 +274,6 @@ gst_omx_audio_dec_change_state (GstElement * element, GstStateChange transition)
     case GST_STATE_CHANGE_PAUSED_TO_READY:
       self->downstream_flow_ret = GST_FLOW_FLUSHING;
       self->started = FALSE;
-
-      if (!gst_omx_audio_dec_shutdown (self))
```
Comment (michaelgruner) [inline omx/gstomxaudiodec.c:283]: Why is this no longer required?

File: omx/gstomxaudiodec.c [hunk 15]
```diff
@@ -397,12 +375,12 @@ gst_omx_audio_dec_loop (GstOMXAudioDec * self)
         default:
           /* This will break the outer loop too as the
            * i == pcm_param.nChannels afterwards */
-          for (i = 0; i < pcm_param.nChannels; i++)
+          for (i = 0; i < self->pcm_param.nChannels; i++)
```
Comment (michaelgruner) [inline omx/gstomxaudiodec.c]: To me, it doesn't make sense to invalidate all the positions if a single one of them is invalid. Should we leave only the `i` position as `GST_AUDIO_CHANNEL_POSITION_NONE`?

File: omx/gstomxaudiodec.c [hunk 16]
```diff
@@ -762,9 +749,6 @@ gst_omx_audio_dec_stop (GstAudioDecoder * decoder)
 
   GST_DEBUG_OBJECT (self, "Stopping decoder");
 
-  gst_omx_port_set_flushing (self->dec_in_port, 5 * GST_SECOND, TRUE);
```
Comment (michaelgruner) [inline omx/gstomxaudiodec.c:765]: Why is this no longer required?

File: omx/gstomxaudiodec.c [hunk 17]
```diff
@@ -867,29 +850,38 @@ gst_omx_audio_dec_set_format (GstAudioDecoder * decoder, GstCaps * caps)
...
       return FALSE;
     }
   }
 
-  GST_DEBUG_OBJECT (self, "Updating outport port definition");
-  if (gst_omx_port_update_port_definition (self->dec_out_port,
-          NULL) != OMX_ErrorNone)
-    return FALSE;
+  /* Call subclass set_format for output port configuration */
+  if (klass->set_format) {
```
Comment (michaelgruner) [inline omx/gstomxaudiodec.c]: This conditional is unnecessary, move it to the `if (klass->set_format)` above

File: omx/gstomxaudiodec.c [hunk 18]
```diff
@@ -917,21 +909,38 @@ gst_omx_audio_dec_set_format (GstAudioDecoder * decoder, GstCaps * caps)
       return FALSE;
   } else {
     if (!(klass->cdata.hacks & GST_OMX_HACK_NO_DISABLE_OUTPORT)) {
-      /* Disable output port */
-      if (gst_omx_port_set_enabled (self->dec_out_port, FALSE) != OMX_ErrorNone)
+
+      GST_DEBUG_OBJECT (self, "Enabling ports");
+
+      /* Enable input port */
+      if (gst_omx_port_set_enabled (self->dec_in_port, TRUE) != OMX_ErrorNone)
```
Comment (michaelgruner) [inline omx/gstomxaudiodec.c:913]: If you are not returning an error code, and if the method can fail in several points, I recommend adding error messages.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

File: omx/gstomxaacdec.c (hunk 10 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: Yes, nice observation. Ready: ac664c134ac783d37aec797017d18a2765849fe3

File: omx/gstomxaacdec.c (hunk 12 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: @michaelgruner That applies only for the omx input port configuration, for the omx output port configuration they are not necessary.

File: omx/gstomxaudiodec.c (hunk 13 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:255]: This is part of the element tear down fix. Flushing the component at this points causes a omx component state corruption. Removing this allows the component to close properly as well as the gstreamer element.

File: omx/gstomxaudiodec.c (hunk 14 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:283]: This is call was a duplicate, the class has the _close call that gets in charge of calling the _shutdown function. The _close function is already registered in the _class_init. This is also part of an element tear down fix.

File: omx/gstomxaudiodec.c (hunk 15 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c]: I agree with you. 37e7289e350de4489759ea0bee787ef86819ab60

File: omx/gstomxaudiodec.c (hunk 16 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:765]: Omx component and element tear down fix (as explained in comment from above)

File: omx/gstomxaudiodec.c (hunk 18 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:913]: Checking the methods you pointed out,  they are already using the GST_DEBUG_ resource to report errors and warnings, this is done internally in the gstomx core file. So adding extra external may add unnecessary noise.

File: omx/gstomxaudiodec.c (hunk 17 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c]: Removed: de21a4ad7bcb37241fc66178972f27febe8b83f5

**Comment by joselopez10** [comment]
@michaelgruner you can now find requested changes applied plus comments. - Thanks

File: omx/gstomxaacdec.c (hunk 12 above)
Comment (michaelgruner) [reply omx/gstomxaacdec.c]: I don't understand. So, is it possible for an input port to don't have caps? is that a normal case and not an error?

**Review by michaelgruner — CHANGES_REQUESTED** [review]

File: omx/gstomxaacdec.c (hunk 12 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: @michaelgruner Sorry my answer was not clear.  >> Q: If caps == NULL shouldn't you return an error? >> A: Yes you are correct. I removed that condition and placed the proper pointer guards for caps/pointer checking in the class methods. 80f12f3b53c13492fb3b54d22b206df13695b772

File: omx/gstomxaudiodec.c (hunk 13 above)
Comment (cgomezv) [reply omx/gstomxaudiodec.c:255]: Yes, it seems that the audio codecs don't support flushing ports, but I think it's better to indicate that at this moment the ports are in flushing state and no more data should be processed, so I would add instead the following lines:

```
      /* DSP does not support flush ports */
      self->dec_out_port->flushing = TRUE;
      self->dec_in_port->flushing = TRUE;
```

**Review by cgomezv — CHANGES_REQUESTED** [review]

File: omx/gstomxaudiodec.c (hunk 13 above)
Comment (joselopez10) [reply omx/gstomxaudiodec.c:255]: Addressed: 85952ce66a4ccdd9e628f924c0d678b33934f7c5

**Review by cgomezv — APPROVED** [review]

File: omx/gstomxaacdec.c [hunk 19]
```diff
@@ -79,87 +78,125 @@ gst_omx_aac_dec_class_init (GstOMXAACDecClass * klass)
 static void
 gst_omx_aac_dec_init (GstOMXAACDec * self)
 {
-  /* FIXME: Other values exist too! */
-  self->spf = 1024;
+  g_return_val_if_fail (self, NULL);
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: This should give you a warning? Does this compile?

File: omx/gstomxaacdec.c [hunk 20]
```diff
@@ -79,87 +78,125 @@ gst_omx_aac_dec_class_init (GstOMXAACDecClass * klass)
...
-  GstOMXAACDec *self = GST_OMX_AAC_DEC (dec);
+  GstOMXAACDec *self = NULL;
   OMX_PARAM_PORTDEFINITIONTYPE port_def;
   OMX_AUDIO_PARAM_AACPROFILETYPE aac_param;
   OMX_ERRORTYPE err;
   GstStructure *s;
   gint rate, channels, mpegversion;
   const gchar *stream_format;
 
+  g_return_val_if_fail (dec, NULL);
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: This should give you a warning, does this compile? This function returns `gboolean`, you are trying to return `NULL`

File: omx/gstomxaacdec.c [hunk 21]
```diff
@@ -169,13 +206,19 @@ static gboolean
...
 {
-  GstOMXAACDec *self = GST_OMX_AAC_DEC (dec);
+  GstOMXAACDec *self = NULL;
   OMX_AUDIO_PARAM_AACPROFILETYPE aac_param;
   OMX_ERRORTYPE err;
   GstStructure *s;
   gint rate, channels, mpegversion;
   const gchar *stream_format;
 
+  g_return_val_if_fail (dec, NULL);
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: Returning `NULL` in a `gboolean` method

File: omx/gstomxaacdec.c [hunk 22]
```diff
@@ -232,6 +275,7 @@ gst_omx_aac_dec_is_format_change (GstOMXAudioDec * dec, GstOMXPort * port,
 static gint
 gst_omx_aac_dec_get_samples_per_frame (GstOMXAudioDec * dec, GstOMXPort * port)
 {
+  g_return_val_if_fail (dec, NULL);
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: Returning `NULL` in a `gint` method

File: omx/gstomxaacdec.c [hunk 23]
```diff
@@ -242,6 +286,9 @@ gst_omx_aac_dec_get_channel_positions (GstOMXAudioDec * dec,
   OMX_AUDIO_PARAM_PCMMODETYPE pcm_param;
   OMX_ERRORTYPE err;
 
+  g_return_val_if_fail (dec, NULL);
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: Returning `NULL` in a `gboolean` method

**Review by michaelgruner — CHANGES_REQUESTED** [review]
Please check you build for warnings. There are a lot of syntax problems

File: omx/gstomxaacdec.c (hunk 19 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: Yes it compiled that way, didn't catch the warnings at build time. Here 0947ed4 the g_return_val_if_fail function incorrect usage is addressed. Please review

File: omx/gstomxaacdec.c (hunk 20 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: Ready for review: 0947ed4a695b58965705a115cfebaebbfdd2f0db

File: omx/gstomxaacdec.c (hunk 21 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: Ready for review: 0947ed4a695b58965705a115cfebaebbfdd2f0db

File: omx/gstomxaacdec.c (hunk 22 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: Ready for review: 0947ed4a695b58965705a115cfebaebbfdd2f0db

File: omx/gstomxaacdec.c (hunk 23 above)
Comment (joselopez10) [reply omx/gstomxaacdec.c]: Ready for review: 0947ed4a695b58965705a115cfebaebbfdd2f0db

File: omx/gstomxaacdec.c [hunk 24]
```diff
@@ -232,6 +275,7 @@ gst_omx_aac_dec_is_format_change (GstOMXAudioDec * dec, GstOMXPort * port,
 static gint
 gst_omx_aac_dec_get_samples_per_frame (GstOMXAudioDec * dec, GstOMXPort * port)
 {
+  g_return_val_if_fail (dec, GST_FLOW_ERROR);
```
Comment (michaelgruner) [inline omx/gstomxaacdec.c]: `GST_FLOW_ERROR` has no special meaning in this context. The receiver will interpret it as a valid integer and use it. I recommend an invalid value, such as -1 or 0.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

**Review by michaelgruner — APPROVED** [review]

**Pull request merged**